msim/src/sim/net/network.rs

```diff
@@ -226,46 +274,88 @@ impl Network {
 
     pub fn send(
         &mut self,
-        node: NodeId,
+        node_id: NodeId,
         src: SocketAddr,
         dst: SocketAddr,
         tag: u64,
         data: Payload,
-    ) {
-        trace!("send: {node} {src} -> {dst}, tag={tag:x}");
+    ) -> io::Result<()> {
+        trace!("send: {node_id} {src} -> {dst}, tag={tag:x}");
         let dst_node = if dst.ip().is_loopback() {
-            node
+            node_id
         } else if let Some(x) = self.addr_to_node.get(&dst.ip()) {
             *x
         } else {
             trace!("destination not found: {dst}");
-            return;
+            return Err(io::Error::new(
+                io::ErrorKind::ConnectionRefused,
+                format!("host unreachable: {dst}"),
+            ));
         };
-        if self.clogged_node.contains(&node)
+        if self.clogged_node.contains(&node_id)
             || self.clogged_node.contains(&dst_node)
-            || self.clogged_link.contains(&(node, dst_node))
+            || self.clogged_link.contains(&(node_id, dst_node))
         {
             trace!("clogged");
-            return;
+            return Err(io::Error::new(
+                io::ErrorKind::ConnectionRefused,
+                format!("host unreachable: {dst}"),
+            ));
         }
 
-        // TODO: need to distinguish between tcp/udp for the purposes of packet loss.
-        let plr = self
-            .config
-            .packet_loss
-            .packet_loss_rate(&mut self.rand, node, dst_node);
-        if self.rand.gen_bool(plr) {
-            trace!("packet loss");
-            return;
+        match data.ty {
+            PayloadType::Udp => {
+                let plr =
+                    self.config
+                        .packet_loss
+                        .packet_loss_rate(&mut self.rand, node_id, dst_node);
+                if self.rand.gen_bool(plr) {
+                    trace!("packet loss");
+                    return Ok(());
+                }
+            }
+            PayloadType::TcpSignalConnect | PayloadType::TcpData => {
+                let plr =
+                    self.config
+                        .tcp_packet_loss
+                        .packet_loss_rate(&mut self.rand, node_id, dst_node);
+                if self.rand.gen_bool(plr) {
+                    trace!("tcp connection failure");
+                    return Err(io::Error::new(
+                        io::ErrorKind::ConnectionReset,
+                        format!("peer hung up: {dst}"),
+                    ));
```

Comment:
A bit confused here. In tcp, if a single packet is lost, it is retried. Why are we simulating a failed connection for `TcpData` case? I assume `TcpSignalConnect` is for tcp handshake so I suppose that part makes sense to me

Reply:
We don't simulate tcp at the protocol level (with retransmits, window sizes, etc), only at the application level (reliable, in-order delivery). Therefore individual tcp "segments" must not be dropped. All we can simulate are application-level phenomena such as latency and connection failures.